### ISSUE TITLE
Provide a command line interface (CLI)

### ISSUE THREAD

**Issue by tkrotoff** [cgiffard/node-simplecrawler#185]
See https://github.com/cgiffard/node-simplecrawler/blob/v0.5.4/lib/cli.js

**Comment by cgiffard** [comment]
I'll do what I can! Always wanted to have one. :)

**Comment by cgiffard** [comment]
FYI working on this now.

**Comment by tmpfs** [comment]
I am working on this right now and would like to start a conversation with you about it.

I started out to fulfill two distinct requirements:

* Validate all HTML5 and XHTML pages in a website with one command
* Dynamically generate a sitemap from the HTML `<meta>` elements

But as I am learning `simplecrawler` I see so much potential, if you would like to take a look at the current code it is here:

https://github.com/tmpfs/linkdown

Validation is implemented and working, and now you guys have helped me with the *same domain* problem I will get on with printing a site map tree (then later I will convert to markdown -> html).

I am considering implementing the following commands also:

* `ls`: Print links for each url (I think I need `discoverResources` for this)
* `get`: Download all or part of a website to a cache implementation (initially filesystem but would be cool to stash in a database too, maybe `couchdb` for querying offline?)

And have ideas for more. I want to implement `--report` to save a JSON report to disc and `--summary` to print the report at the end of a crawl, `--stats` to print the statistics you guys so kindly expose :)

I would love your feedback at this early stage so that we could discuss the option implementations for exclude paths (fetch conditions), exclude MIME types etc etc.

If you don't want to install java/nu validator run the `info` command and play around with `--log-level` to get a feel for where I am going with this.

**Comment by cgiffard** [comment]
I feel like there's definitely room for your CLI — since it definitely provides a distinct and useful set of functionality — and the default CLI to exist concurrently.

I'm planning a flexible arrangement with different `reporters` — ways of formatting/outputting data — and an arrangement that uses the node REPL to give you an interface to the crawler as it's running.

I've been busy in the day job and haven't made much progress in the last little while, but I've pushed it to the remote branch `feature/cli-tool` if you're curious to see it.

**Comment by tmpfs** [comment]
Like it :) Will definitely be checking the source for the progress indicator!

I see what you mean, yours is a low-level interface to the inner workings of `simplecrawler` whilst I am aiming for higher level commands that perform common tasks.

BTW, need to add `colors` to your deps.

**Comment by cgiffard** [comment]
Of course, there's a lot of basic code hygiene that simply hasn't been done to that PR yet. Consider it a rough sketch, not a merge-ready feature. :)

**Comment by tmpfs** [comment]
FYI, `linkdown@1.0.10` now supports pausing and resuming crawls using the `TSTP` and `CONT` signals respectively (can't use `STOP` unfortunately).

Works like a charm, thanks @cgiffard and @fredrikekelund :)

**Comment by tmpfs** [comment]
For the record:

https://github.com/tmpfs/linkdown#credits

:)

Can now also `exec` commands and I am extracting meta data from the HTML pages, should soon have a `tree` command implemented and I will be very close to my goal of automatic sitemap generation :)
